<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using IntegratedLearner for multi-omics prediction and classification • IntegratedLearner</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using IntegratedLearner for multi-omics prediction and classification">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">IntegratedLearner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/IntegratedLearner.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/himelmallick/IntegratedLearner/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using IntegratedLearner for multi-omics
prediction and classification</h1>
                        <h4 data-toc-skip class="author">Himel Mallick
and Anupreet Porwal</h4>
            
            <h4 data-toc-skip class="date">2024-08-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/himelmallick/IntegratedLearner/blob/master/vignettes/IntegratedLearner.Rmd" class="external-link"><code>vignettes/IntegratedLearner.Rmd</code></a></small>
      <div class="hidden name"><code>IntegratedLearner.Rmd</code></div>

    </div>

    
    
<p>This vignette highlights some example workflows for performing
multi-omics prediction and classification using the
<code>IntegratedLearner</code> R package.</p>
<p><code>IntegratedLearner</code> provides an integrated machine
learning framework to 1) consolidate predictions by borrowing
information across several longitudinal and cross-sectional omics data
layers, 2) decipher the mechanistic role of individual omics features
that can potentially lead to new sets of testable hypotheses, and 3)
quantify uncertainty of the integration process. Three types of
integration paradigms are supported: early, late, and intermediate. The
software includes multiple ML models based on the <a href="https://cran.r-project.org/web/packages/SuperLearner/index.html" class="external-link">SuperLearner
R package</a> as well as several data exploration capabilities and
visualization modules in a unified estimation framework.</p>
<div class="section level2">
<h2 id="load-packages">Load Packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<p>Once installed, <strong><code>IntegratedLearner</code></strong> can
be simply loaded (along with the required packages) as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the IntegratedLearner package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/himelmallick/IntegratedLearner" class="external-link">IntegratedLearner</a></span><span class="op">)</span></span>
<span><span class="co"># This is needed for running BART</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>java.parameters <span class="op">=</span> <span class="st">"-Xmx6G"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load other essential packages for this tutorial</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ecpolley/SuperLearner" class="external-link">SuperLearner</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/topepo/caret/" class="external-link">caret</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/bayesplot/" class="external-link">bayesplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bartMachine</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-input">The Input<a class="anchor" aria-label="anchor" href="#the-input"></a>
</h2>
<p><strong><code>IntegratedLearner</code></strong> requires three
tab-delimited input files: (i) concatenated multi-omics profiles
(<code>feature_table</code>), (ii) sample-specific metadata
(<code>sample_metadata</code>), and (iii) feature-specific metadata
(<code>feature_metadata</code>). The rows of the
<code>feature_table</code> should correspond to the concatenated
features (e.g., microbiome, gene expression, metabolites, etc.) and the
columns should correspond to samples.</p>
<p>The columns of the <code>sample_metadata</code> must correspond to
sample-specific covariates (e.g., disease status or the clinical outcome
of interest) with the rows corresponding to samples. Row names of
<code>sample_metadata</code> must match the column names of
<code>feature_table</code>. Furthermore, <code>sample_metadata</code>
should have a column named <strong><code>subjectID</code></strong>
describing per-subject unique identifiers. For longitudinal designs,
this variable is expected to have non-unique values. Additionally, a
column named <strong><code>Y</code></strong> must be present which is
the outcome of interest (can be binary or continuous).</p>
<p><code>feature_metadata</code> is expected to be a data frame
containing feature-specific metadata with a column named
<strong><code>featureID</code></strong> describing per-feature unique
identifiers and <strong><code>featureType</code></strong> describing the
corresponding omics source layers (e.g., metagenomics, metabolomics,
etc.). Row names of <code>feature_metadata</code> must match that of
<code>feature_table</code>.</p>
<p>For the purpose of this vignette, it is assumed that these three
input data have already been quality-controlled with necessary
preprocessing steps. For the classification example, we will be using a
cleaned version of the PRISM multi-omics dataset (<a href="https://www.nature.com/articles/s41564-018-0306-4" class="external-link">Franzosa et
al., 2019</a>) which can be downloaded from <a href="https://github.com/himelmallick/IntegratedLearner/blob/master/data/PRISM.RData" class="external-link">here</a>.</p>
<div class="section level3">
<h3 id="example-1---prediction-of-binary-ibd-disease-status-from-multi-omics-profiles">Example 1 - Prediction of Binary IBD Disease Status from Multi-omics
Profiles<a class="anchor" aria-label="anchor" href="#example-1---prediction-of-binary-ibd-disease-status-from-multi-omics-profiles"></a>
</h3>
<p>We first use the PRISM dataset (<a href="https://www.nature.com/articles/s41564-018-0306-4" class="external-link">Franzosa et
al., 2019</a>) which is a gut microbiome multi-omics dataset consisting
of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>9171</mn><annotation encoding="application/x-tex">9171</annotation></semantics></math>
quality-controlled features from 2 layers (i.e., microbiome taxonomic
profiles and metabolites). In this study, stool samples were collected
from a cross-sectional cohort of individuals enrolled in the Prospective
Registry in IBD Study at MGH (PRISM) in order to characterize the gut
metabolic profile and microbiome composition in Inflammatory Bowel
Diseases (IBD).</p>
<p>This cohort included 155 subjects: 68 with Crohn’s disease (CD), 53
with ulcerative colitis (UC), jointly grouped as IBD, and 34 non-IBD
controls. Each stool sample was subjected to metagenomic sequencing
followed by profiling of microbial community taxonomic composition and
functional potential. In addition, each sample was analyzed by four
liquid chromatography tandem mass spectrometry (LC-MS) methods measuring
polar metabolites, lipids, free fatty acids, and bile acids,
respectively.</p>
<p>In addition to carrying out a holistic investigation of the
microbiome–metabolome interface, one of the primary objectives of this
study was to assess the power of the metabolomic and microbial layers in
classifying IBD status.</p>
<p>Let us first examine the characteristics of the PRISM data by loading
the <a href="https://github.com/himelmallick/IntegratedLearner/blob/master/data/PRISM.RData" class="external-link"><strong><code>PRISM.RData</code></strong></a>
object which contains a list of three data frames:
<code>feature_table</code>, <code>sample_metadata</code> and
<code>feature_metadata</code>. Note that, the <code>feature_table</code>
contains negative values. This is because both species and metabolite
data have been residualized to remove the effect of potential
confounders.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Load dataset </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://github.com/himelmallick/IntegratedLearner/blob/master/data/PRISM.RData?raw=true"</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># Extract individual components </span></span>
<span><span class="va">feature_table</span><span class="op">&lt;-</span><span class="va">pcl</span><span class="op">$</span><span class="va">feature_table</span></span>
<span><span class="va">sample_metadata</span><span class="op">&lt;-</span><span class="va">pcl</span><span class="op">$</span><span class="va">sample_metadata</span></span>
<span><span class="va">feature_metadata</span><span class="op">&lt;-</span><span class="va">pcl</span><span class="op">$</span><span class="va">feature_metadata</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">pcl</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># Explore data dimensions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">feature_table</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                G35127      G35128      G35152       G36347</span></span>
<span><span class="co">#&gt; Granulicella_unclassified -0.05253649 -0.05127158 -0.06133085  0.004887447</span></span>
<span><span class="co">#&gt; Actinomyces_graevenitzii   1.04668500 -1.32629194 -1.51654615 -3.247989324</span></span>
<span><span class="co">#&gt; Actinomyces_johnsonii     -0.70327678 -0.41575776 -0.29326475 -0.314361595</span></span>
<span><span class="co">#&gt; Actinomyces_massiliensis  -0.56808952  0.14722099  0.05660884 -1.077235688</span></span>
<span><span class="co">#&gt; Actinomyces_naeslundii    -0.49546119 -0.15921604 -0.03146485 -0.354377267</span></span>
<span><span class="co">#&gt;                                 G36348</span></span>
<span><span class="co">#&gt; Granulicella_unclassified -0.006164066</span></span>
<span><span class="co">#&gt; Actinomyces_graevenitzii  -0.717183019</span></span>
<span><span class="co">#&gt; Actinomyces_johnsonii     -0.340485318</span></span>
<span><span class="co">#&gt; Actinomyces_massiliensis  -0.159240362</span></span>
<span><span class="co">#&gt; Actinomyces_naeslundii    -0.139758576</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;        Diagnosis Y subjectID</span></span>
<span><span class="co">#&gt; G35127        CD 1    G35127</span></span>
<span><span class="co">#&gt; G35128        CD 1    G35128</span></span>
<span><span class="co">#&gt; G35152        CD 1    G35152</span></span>
<span><span class="co">#&gt; G36347        CD 1    G36347</span></span>
<span><span class="co">#&gt; G36348        CD 1    G36348</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">feature_metadata</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                           featureID featureType</span></span>
<span><span class="co">#&gt; Granulicella_unclassified Granulicella_unclassified     species</span></span>
<span><span class="co">#&gt; Actinomyces_graevenitzii   Actinomyces_graevenitzii     species</span></span>
<span><span class="co">#&gt; Actinomyces_johnsonii         Actinomyces_johnsonii     species</span></span>
<span><span class="co">#&gt; Actinomyces_massiliensis   Actinomyces_massiliensis     species</span></span>
<span><span class="co">#&gt; Actinomyces_naeslundii       Actinomyces_naeslundii     species</span></span>
<span>  </span>
<span><span class="co"># How many layers and how many features per layer?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">feature_metadata</span><span class="op">$</span><span class="va">featureType</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; metabolites     species </span></span>
<span><span class="co">#&gt;        8831         340</span></span>
<span>  </span>
<span><span class="co"># Distribution of outcome (1: IBD, 0: nonIBD)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">$</span><span class="va">Y</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   0   1 </span></span>
<span><span class="co">#&gt;  34 121</span></span>
<span>  </span>
<span><span class="co"># Sanity check</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature_table</span><span class="op">)</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature_metadata</span><span class="op">)</span><span class="op">)</span> <span class="co"># TRUE</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">feature_table</span><span class="op">)</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">)</span><span class="op">)</span> <span class="co"># TRUE</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="co"># Load independent validation dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://github.com/himelmallick/IntegratedLearner/blob/master/data/NLIBD.RData?raw=true"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">feature_table_valid</span><span class="op">&lt;-</span><span class="va">pcl</span><span class="op">$</span><span class="va">feature_table</span></span>
<span><span class="va">sample_metadata_valid</span><span class="op">&lt;-</span><span class="va">pcl</span><span class="op">$</span><span class="va">sample_metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">pcl</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Sanity check to make sure test set has sample structure as training </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature_table</span><span class="op">)</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature_table_valid</span><span class="op">)</span><span class="op">)</span> <span class="co"># TRUE</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">feature_table_valid</span><span class="op">)</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sample_metadata_valid</span><span class="op">)</span><span class="op">)</span> <span class="co"># TRUE</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p><code>IntegratedLearner</code> late fusion algorithm proceeds by 1)
fitting a machine learning algorithm per-layer to predict outcome
(<code>base_learner</code>) and 2) combining the layer-wise
cross-validated predictions using a meta model
(<code>meta_learner</code>) to generate final predictions based on all
available data points. As a default choice, we recommend
<code>SL.nnls.auc</code> as the meta model algorithm. It fits a
non-negative least squares (in case of continuous outcome) and rank loss
minimization (in case of binary outcome) on layer-wise cross-validated
predictions to generate the final predictions and quantify per-layer
contribution in the final predictions. As an example, we would like to
build a random forest classifier based on these data to classify IBD
patients. By default, <code>IntegratedLearner</code> uses a 5-fold CV to
train the model (for the full dataset, it takes about ~5-6 minutes using
a single core of a system with an Intel Core i5 processor (1.7 GHz) and
16 GB of RAM - adjust your expectations accordingly!).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">fit</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/IntegratedLearner.html">IntegratedLearner</a></span><span class="op">(</span>feature_table <span class="op">=</span> <span class="va">feature_table</span>,</span>
<span>                               sample_metadata <span class="op">=</span> <span class="va">sample_metadata</span>, </span>
<span>                               feature_metadata <span class="op">=</span> <span class="va">feature_metadata</span>,</span>
<span>                               feature_table_valid <span class="op">=</span> <span class="va">feature_table_valid</span>,</span>
<span>                               sample_metadata_valid <span class="op">=</span> <span class="va">sample_metadata_valid</span>,</span>
<span>                               folds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                               base_learner <span class="op">=</span> <span class="st">'SL.randomForest'</span>,</span>
<span>                               meta_learner <span class="op">=</span> <span class="st">'SL.nnls.auc'</span>,</span>
<span>                               verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                               family<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/family.html" class="external-link">binomial</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Running base model for layer  1 ...</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 8831</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 8831</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 8831</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 8831</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 8831</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Non-Negative least squares convergence: TRUE</span></span>
<span><span class="co">#&gt; full SL.randomForest_All</span></span>
<span><span class="co">#&gt; Running base model for layer  2 ...</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 340</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 340</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 340</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 340</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 340</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Non-Negative least squares convergence: TRUE</span></span>
<span><span class="co">#&gt; full SL.randomForest_All</span></span>
<span><span class="co">#&gt; Running stacked model...</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 2</span></span>
<span><span class="co">#&gt; CV SL.nnls.auc_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 2</span></span>
<span><span class="co">#&gt; CV SL.nnls.auc_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 2</span></span>
<span><span class="co">#&gt; CV SL.nnls.auc_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 2</span></span>
<span><span class="co">#&gt; CV SL.nnls.auc_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 2</span></span>
<span><span class="co">#&gt; CV SL.nnls.auc_All</span></span>
<span><span class="co">#&gt; Non-Negative least squares convergence: TRUE</span></span>
<span><span class="co">#&gt; full SL.nnls.auc_All</span></span>
<span><span class="co">#&gt; Running concatenated model...</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 9171</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 9171</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 9171</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 9171</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Number of covariates in All is: 9171</span></span>
<span><span class="co">#&gt; CV SL.randomForest_All</span></span>
<span><span class="co">#&gt; Non-Negative least squares convergence: TRUE</span></span>
<span><span class="co">#&gt; full SL.randomForest_All</span></span>
<span><span class="co">#&gt; Time for model fit : 3.762 minutes </span></span>
<span><span class="co">#&gt; ========================================</span></span>
<span><span class="co">#&gt; Model fit for individual layers: SL.randomForest </span></span>
<span><span class="co">#&gt; Model fit for stacked layer: SL.nnls.auc </span></span>
<span><span class="co">#&gt; Model fit for concatenated layer: SL.randomForest </span></span>
<span><span class="co">#&gt; ========================================</span></span>
<span><span class="co">#&gt; AUC metric for training data: </span></span>
<span><span class="co">#&gt; Individual layers: </span></span>
<span><span class="co">#&gt; metabolites     species </span></span>
<span><span class="co">#&gt;       0.934       0.955 </span></span>
<span><span class="co">#&gt; ======================</span></span>
<span><span class="co">#&gt; Stacked model:0.951 </span></span>
<span><span class="co">#&gt; ======================</span></span>
<span><span class="co">#&gt; Concatenated model:0.957 </span></span>
<span><span class="co">#&gt; ======================</span></span>
<span><span class="co">#&gt; ========================================</span></span>
<span><span class="co">#&gt; AUC metric for test data: </span></span>
<span><span class="co">#&gt; Individual layers: </span></span>
<span><span class="co">#&gt; metabolites     species </span></span>
<span><span class="co">#&gt;       0.782       0.879 </span></span>
<span><span class="co">#&gt; ======================</span></span>
<span><span class="co">#&gt; Stacked model:0.908 </span></span>
<span><span class="co">#&gt; ======================</span></span>
<span><span class="co">#&gt; Concatenated model:0.895 </span></span>
<span><span class="co">#&gt; ======================</span></span>
<span><span class="co">#&gt; ========================================</span></span>
<span><span class="co">#&gt; Weights for individual layers predictions in IntegratedLearner: </span></span>
<span><span class="co">#&gt; metabolites     species </span></span>
<span><span class="co">#&gt;        0.18        0.82 </span></span>
<span><span class="co">#&gt; ========================================</span></span></code></pre></div>
<p><code>IntegratedLearner</code> offers easily accessible and
interpretable summary outputs including 1) computation time, 2)
per-layer AUC/
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>R</mi><mn>2</mn></msup><annotation encoding="application/x-tex">R^2</annotation></semantics></math>
scores on training and/or test data (if <code>feature_table_valid</code>
and <code>sample_metadata_valid</code> are provided), 3) AUC/
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>R</mi><mn>2</mn></msup><annotation encoding="application/x-tex">R^2</annotation></semantics></math>
metrics for stacked and concatenated model if
<code>run_stacked=TRUE</code> and <code>run_concat=TRUE</code> and 4)
estimated per-layer weights from meta learner in stacked model.</p>
<p>We can visualize the classification performance by constructing
layer-wise ROC curves for both train and test set using
<code><a href="../reference/plot.learner.html">plot.learner()</a></code> function that takes
<code>IntegratedLearner</code> object as input:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot.obj</span> <span class="op">&lt;-</span> <span class="fu">IntegratedLearner</span><span class="fu">:::</span><span class="fu"><a href="../reference/plot.learner.html">plot.learner</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntegratedLearner_files/figure-html/unnamed-chunk-4-1.png" class="r-plt" width="672"></p>
<p>For this particular multi-omics dataset, species data is more
predictive than metabolites in classifying IBD patients and the stacked
model achieves superior accuracy to individual layers and concatenation
model in independent validation, indicating that the stacked multi-omics
classifier leads to a competitive or superior cross-validated and
independent validation classification accuracy than its single-omics
counterparts.</p>
</div>
<div class="section level3">
<h3 id="example-2---prediction-of-continuous-gestational-age-from-multi-omics-profiles">Example 2 - Prediction of Continuous Gestational Age from
Multi-omics Profiles<a class="anchor" aria-label="anchor" href="#example-2---prediction-of-continuous-gestational-age-from-multi-omics-profiles"></a>
</h3>
<p>The second dataset is a longitudinal multi-omics data from pregnant
women in a cohort study at Stanford University (<a href="https://academic.oup.com/bioinformatics/article/35/1/95/5047759" class="external-link">Ghaemi
et al., 2019</a>) that aimed to prospectively examine environmental and
biological factors associated with normal and pathological pregnancies.
Women were eligible if they were at least 18 years of age and in their
first trimester of a singleton pregnancy. Unlike the PRISM dataset, the
outcome variable in this study (gestational age) is a continuous
outcome, which was determined by best obstetrical estimate as
recommended by the American College of Obstetricians and Gynecologists
.</p>
<p>In 17 women, three samples were collected during pregnancy and a
fourth one after delivery. The time points were chosen such that a
peripheral blood sample (CyTOF analysis), a plasma sample (proteomic,
cell-free transcriptomics, metabolomics analyses), a serum sample
(luminex analyses) and a series of culture swabs (microbiome analysis)
were simultaneously collected from each woman during the first (7–14
weeks), second (15–20 weeks) and third (24–32 weeks) trimester of
pregnancy and 6-week postpartum.</p>
<p>In order to assess performance of various machine learning modules
available in <code>IntegratedLearner</code>, we calculate the
coefficient of determination
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>R</mi><mn>2</mn></msup><annotation encoding="application/x-tex">R^2</annotation></semantics></math>)
based on the observed and cross-validated out-of-sample predicted values
of the gestational age. As before, cross-validation folds are
synchronized between the individual base models from each dataset to
leave out the same set of data points at all levels of the analysis.</p>
<p>As before, let’s now examine the characteristics of the pregnancy
data by loading the <a href="https://github.com/himelmallick/IntegratedLearner/blob/master/data/pregnancy.RData" class="external-link"><strong><code>pregnancy.RData</code></strong></a>
object which again contains a list of three data frames:
<code>feature_table</code>, <code>sample_metadata</code> and
<code>feature_metadata</code>. Unlike the PRISM study, this dataset
contains repeated measures during pregnancy that allowed assessing
important biological adaptations occurring continuously from the early
phases of fetal development (first trimester) to the late phases of
gestation (third trimester). Considering both the small sample size and
the repeated measures aspects of this study, we employ a
one-subject-leave-out cross-validation to build prediction models,
following the original study.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Load dataset </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://github.com/himelmallick/IntegratedLearner/blob/master/data/pregnancy.RData?raw=true"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract individual components </span></span>
<span><span class="va">feature_table</span><span class="op">&lt;-</span><span class="va">pcl</span><span class="op">$</span><span class="va">feature_table</span></span>
<span><span class="va">sample_metadata</span><span class="op">&lt;-</span><span class="va">pcl</span><span class="op">$</span><span class="va">sample_metadata</span></span>
<span><span class="va">feature_metadata</span><span class="op">&lt;-</span><span class="va">pcl</span><span class="op">$</span><span class="va">feature_metadata</span></span>
<span>  </span>
<span><span class="co"># Explore data dimensions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">feature_table</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;        PTLG002_1 PTLG003_1  PTLG004_1 PTLG005_1 PTLG007_1</span></span>
<span><span class="co">#&gt; CEP135  28.21785  54.56723  53.776824  15.26909  11.04831</span></span>
<span><span class="co">#&gt; MIIP    10.10756  17.11006   4.336841   0.00000  19.88695</span></span>
<span><span class="co">#&gt; GNL3    45.25968  58.26670  56.378929  70.23780  64.08018</span></span>
<span><span class="co">#&gt; CEP70   79.09550  67.97782  93.675759 128.26033  66.28985</span></span>
<span><span class="co">#&gt; TIMP1  172.23675 121.62018 183.014677 247.35921 304.93329</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;            Y subjectID</span></span>
<span><span class="co">#&gt; PTLG002_1 11   PTLG002</span></span>
<span><span class="co">#&gt; PTLG003_1 11   PTLG003</span></span>
<span><span class="co">#&gt; PTLG004_1 11   PTLG004</span></span>
<span><span class="co">#&gt; PTLG005_1 11   PTLG005</span></span>
<span><span class="co">#&gt; PTLG007_1 11   PTLG007</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">feature_metadata</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;        featureID featureType</span></span>
<span><span class="co">#&gt; CEP135    CEP135 CellfreeRNA</span></span>
<span><span class="co">#&gt; MIIP        MIIP CellfreeRNA</span></span>
<span><span class="co">#&gt; GNL3        GNL3 CellfreeRNA</span></span>
<span><span class="co">#&gt; CEP70      CEP70 CellfreeRNA</span></span>
<span><span class="co">#&gt; TIMP1      TIMP1 CellfreeRNA</span></span>
<span>  </span>
<span><span class="co"># How many layers and how many features per layer?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">feature_metadata</span><span class="op">$</span><span class="va">featureType</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     CellfreeRNA    ImmuneSystem    Metabolomics      Microbiome   PlasmaLuminex </span></span>
<span><span class="co">#&gt;            9084             264             253             259              31 </span></span>
<span><span class="co">#&gt; PlasmaSomalogic    SerumLuminex </span></span>
<span><span class="co">#&gt;             650              31</span></span>
<span>  </span>
<span><span class="co"># Number of subjects</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">$</span><span class="va">subjectID</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 17</span></span>
<span>  </span>
<span><span class="co"># Sanity check</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature_table</span><span class="op">)</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature_metadata</span><span class="op">)</span><span class="op">)</span> <span class="co"># TRUE</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">feature_table</span><span class="op">)</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">)</span><span class="op">)</span> <span class="co"># TRUE</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="co"># Subset to a few rows to save computing time</span></span>
<span><span class="co"># top_n&lt;-50</span></span>
<span><span class="co"># subsetIDs&lt;-c(1:top_n, (nrow(feature_table)-top_n+1):nrow(feature_table))</span></span>
<span><span class="co"># feature_table&lt;-feature_table[subsetIDs,]</span></span>
<span><span class="co"># feature_metadata&lt;-feature_metadata[subsetIDs,]</span></span></code></pre></div>
<p>The default base model recommendation for
<code>IntegratedLearner</code> is Bayesian additive regression trees or
BART (<code>base_learner='SL.BART'</code>). Using BART as base-model
yields uncertainty estimates (i.e. credible intervals) of the prediction
and model parameters in addition to reporting a small set of
interpretable features for follow-up experiments (i.e. feature
importance scores).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">fit</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/IntegratedLearner.html">IntegratedLearner</a></span><span class="op">(</span>feature_table <span class="op">=</span> <span class="va">feature_table</span>,</span>
<span>                               sample_metadata <span class="op">=</span> <span class="va">sample_metadata</span>, </span>
<span>                               feature_metadata <span class="op">=</span> <span class="va">feature_metadata</span>,</span>
<span>                               folds <span class="op">=</span> <span class="fl">17</span>,</span>
<span>                               base_learner <span class="op">=</span> <span class="st">'SL.BART'</span>,</span>
<span>                               meta_learner <span class="op">=</span> <span class="st">'SL.nnls.auc'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Running base model for layer  1 ... </span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; Running base model for layer  2 ... </span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; Running base model for layer  3 ... </span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; Running base model for layer  4 ... </span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; Running base model for layer  5 ... </span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; Running base model for layer  6 ... </span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; Running base model for layer  7 ... </span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; Running stacked model...</span></span>
<span><span class="co">#&gt; Running concatenated model...</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; serializing in order to be saved for future R sessions...done</span></span>
<span><span class="co">#&gt; Time for model fit : 4.766 minutes </span></span>
<span><span class="co">#&gt; ========================================</span></span>
<span><span class="co">#&gt; Model fit for individual layers: SL.BART </span></span>
<span><span class="co">#&gt; Model fit for stacked layer: SL.nnls.auc </span></span>
<span><span class="co">#&gt; Model fit for concatenated layer: SL.BART </span></span>
<span><span class="co">#&gt; ========================================</span></span>
<span><span class="co">#&gt; R^2 for training data: </span></span>
<span><span class="co">#&gt; Individual layers: </span></span>
<span><span class="co">#&gt;     CellfreeRNA    ImmuneSystem    Metabolomics      Microbiome   PlasmaLuminex </span></span>
<span><span class="co">#&gt;      0.08542212      0.20491045      0.67114346      0.74897747      0.18314107 </span></span>
<span><span class="co">#&gt; PlasmaSomalogic    SerumLuminex </span></span>
<span><span class="co">#&gt;      0.64129255      0.09200745 </span></span>
<span><span class="co">#&gt; ======================</span></span>
<span><span class="co">#&gt; Stacked model:0.8114878 </span></span>
<span><span class="co">#&gt; ======================</span></span>
<span><span class="co">#&gt; Concatenated model:0.687933 </span></span>
<span><span class="co">#&gt; ======================</span></span>
<span><span class="co">#&gt; ========================================</span></span>
<span><span class="co">#&gt; Weights for individual layers predictions in IntegratedLearner: </span></span>
<span><span class="co">#&gt;     CellfreeRNA    ImmuneSystem    Metabolomics      Microbiome   PlasmaLuminex </span></span>
<span><span class="co">#&gt;           0.000           0.000           0.236           0.587           0.000 </span></span>
<span><span class="co">#&gt; PlasmaSomalogic    SerumLuminex </span></span>
<span><span class="co">#&gt;           0.176           0.000 </span></span>
<span><span class="co">#&gt; ========================================</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot.obj</span> <span class="op">&lt;-</span> <span class="fu">IntegratedLearner</span><span class="fu">:::</span><span class="fu"><a href="../reference/plot.learner.html">plot.learner</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntegratedLearner_files/figure-html/unnamed-chunk-7-1.png" class="r-plt" width="672">
As before, the multi-omics stacked prediction model leads to a better
cross-validated accuracy than its single-omics counterparts and
concatenation model.</p>
<p>When <code>base_learner='SL.BART'</code>, in addition to point
predictions, we can also generate 1) credible intervals for all
observations, 2) estimated layer weights in the meta model and 3)
feature importance scores.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="va">fit</span><span class="op">$</span><span class="va">weights</span></span>
<span></span>
<span><span class="va">dataX</span> <span class="op">&lt;-</span> <span class="va">fit</span><span class="op">$</span><span class="va">X_train_layers</span></span>
<span><span class="va">dataY</span> <span class="op">&lt;-</span> <span class="va">fit</span><span class="op">$</span><span class="va">Y_train</span></span>
<span></span>
<span></span>
<span><span class="va">post.samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">weights</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">post.samples</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dataX</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">post.samples</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span><span class="va">post.samples</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bartMachine/man/bart_machine_get_posterior.html" class="external-link">bart_machine_get_posterior</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">model_fits</span><span class="op">$</span><span class="va">model_layers</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">dataX</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">y_hat_posterior_samples</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">weighted.post.samples</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="st">'+'</span>, <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Map</a></span><span class="op">(</span><span class="st">'*'</span>, <span class="va">post.samples</span>, <span class="va">weights</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">weighted.post.samples</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dataX</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dataY</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dataX</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>We show below the 68% and 95% credible intervals obtained from
stacked model for all 51 observations. The filled circle indicates the
posterior median and empty circle indicates the true value of the
observation.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ord_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">weighted.post.samples</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://mc-stan.org/bayesplot/reference/MCMC-intervals.html" class="external-link">mcmc_intervals</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">weighted.post.samples</span><span class="op">)</span>, prob <span class="op">=</span> <span class="fl">0.68</span>, prob_outer <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_y_discrete</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="va">ord_names</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dataY</span><span class="op">[</span><span class="va">ord_names</span><span class="op">]</span>, y <span class="op">=</span> <span class="va">ord_names</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">3</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Gestational age (in months)"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Observations"</span>,</span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">y</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">y</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="IntegratedLearner_files/figure-html/unnamed-chunk-9-1.png" class="r-plt" width="864"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">omicsEye_theme</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="op">)</span> <span class="op">{</span></span>
<span><span class="co"># set default text format based on categorical and length</span></span>
<span>  <span class="va">angle</span> <span class="op">=</span> <span class="fl">45</span></span>
<span>  <span class="va">hjust</span> <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="va">size</span> <span class="op">=</span> <span class="fl">6</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span> <span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    axis.text.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="va">hjust</span>, angle <span class="op">=</span> <span class="va">angle</span><span class="op">)</span>,</span>
<span>    axis.text.y <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>    axis.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>    plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>    plot.subtitle <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>    legend.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span>, face <span class="op">=</span> <span class="st">'bold'</span><span class="op">)</span>,</span>
<span>    legend.text <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span>,</span>
<span>    axis.line <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">'black'</span>, size <span class="op">=</span> <span class="fl">.25</span><span class="op">)</span>,</span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">'black'</span>, size <span class="op">=</span> <span class="fl">.25</span><span class="op">)</span>,</span>
<span>    axis.line.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">'black'</span>, size <span class="op">=</span> <span class="fl">.25</span><span class="op">)</span>,</span>
<span>    axis.line.y <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">'black'</span>, size <span class="op">=</span> <span class="fl">.25</span><span class="op">)</span>,</span>
<span>    panel.border <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    panel.grid.major <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    panel.grid.minor <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">myColtmp</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cornflowerblue"</span>,<span class="st">"darkcyan"</span>,<span class="st">"orchid4"</span>,</span>
<span>            <span class="st">"brown"</span>,<span class="st">"goldenrod4"</span>,<span class="st">"mistyrose4"</span>,<span class="st">"darkgreen"</span>,<span class="st">"purple"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">VIMP_stack</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind.data.frame</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">weights</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">VIMP_stack</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'mean'</span><span class="op">)</span></span>
<span><span class="va">VIMP_stack</span><span class="op">$</span><span class="va">sd</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">VIMP_stack</span><span class="op">$</span><span class="va">type</span><span class="op">&lt;-</span><span class="st">'stack'</span></span>
<span></span>
<span><span class="va">safe_var_importance</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">model</span>, <span class="va">layer_label</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="va">qq</span> <span class="op">&lt;-</span> <span class="fu">bartMachine</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bartMachine/man/investigate_var_importance.html" class="external-link">investigate_var_importance</a></span><span class="op">(</span><span class="va">model</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind.data.frame</a></span><span class="op">(</span><span class="va">qq</span><span class="op">$</span><span class="va">avg_var_props</span>, <span class="va">qq</span><span class="op">$</span><span class="va">sd_var_props</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'mean'</span>, <span class="st">'sd'</span><span class="op">)</span></span>
<span>    <span class="va">df</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="va">layer_label</span></span>
<span>    <span class="va">df</span></span>
<span>  <span class="op">}</span>, error <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/warning.html" class="external-link">warning</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Skipping variable importance for %s: %s"</span>, <span class="va">layer_label</span>, <span class="fu"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">conditionMessage</a></span><span class="op">(</span><span class="va">e</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co"># Returning an empty data frame keeps the downstream plotting code functional</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="op">)</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="op">)</span>, type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">###############</span></span>
<span><span class="co"># Microbiome #</span></span>
<span><span class="co">###############</span></span>
<span><span class="va">VIMP_microbiome</span> <span class="op">&lt;-</span> <span class="fu">safe_var_importance</span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">model_fits</span><span class="op">$</span><span class="va">model_layers</span><span class="op">$</span><span class="va">Microbiome</span>, <span class="st">'Microbiome'</span><span class="op">)</span></span>
<span><span class="co">#&gt; .....</span></span>
<span></span>
<span><span class="co">###############</span></span>
<span><span class="co"># Plasma Somalogic #</span></span>
<span><span class="co">###############</span></span>
<span><span class="va">VIMP_PlasmaSomalogic</span> <span class="op">&lt;-</span> <span class="fu">safe_var_importance</span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">model_fits</span><span class="op">$</span><span class="va">model_layers</span><span class="op">$</span><span class="va">PlasmaSomalogic</span>, <span class="st">'PlasmaSomalogic'</span><span class="op">)</span></span>
<span><span class="co">#&gt; .</span></span>
<span><span class="co">#&gt; Warning in value[[3L]](cond): Skipping variable importance for PlasmaSomalogic:</span></span>
<span><span class="co">#&gt; java.lang.OutOfMemoryError: Java heap space</span></span>
<span></span>
<span><span class="va">VIMP</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind.data.frame</a></span><span class="op">(</span><span class="va">VIMP_stack</span>,</span>
<span>                                     <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">VIMP_microbiome</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>                                     <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">VIMP_PlasmaSomalogic</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">VIMP</span><span class="op">&lt;-</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="va">VIMP</span>, <span class="st">'ID'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p4</span><span class="op">&lt;-</span><span class="va">VIMP</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">type</span> <span class="op">==</span> <span class="st">'stack'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">mean</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">mean</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">ID</span>,<span class="op">-</span><span class="va">mean</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, fill <span class="op">=</span> <span class="st">'darkseagreen'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co">#coord_flip() +</span></span>
<span>  <span class="fu">omicsEye_theme</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'Layer Weights'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `size` argument of `element_line()` is deprecated as of ggplot2 3.4.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `linewidth` argument instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span></span>
<span></span>
<span><span class="va">p5</span><span class="op">&lt;-</span><span class="va">VIMP</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">type</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Microbiome'</span>, <span class="st">'PlasmaSomalogic'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">mean</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>ID <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace_all</a></span><span class="op">(</span><span class="va">ID</span>, <span class="fu"><a href="https://stringr.tidyverse.org/reference/modifiers.html" class="external-link">fixed</a></span><span class="op">(</span><span class="st">"_"</span><span class="op">)</span>, <span class="st">" "</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">type</span>,</span>
<span>                       levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Microbiome'</span>, <span class="st">'PlasmaSomalogic'</span><span class="op">)</span>,</span>
<span>                       labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Microbiome'</span>, <span class="st">'PlasmaSomalogic'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">ID</span>, <span class="op">-</span><span class="va">mean</span><span class="op">)</span>, <span class="va">mean</span>, fill <span class="op">=</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="va">.</span><span class="op">~</span> <span class="va">type</span>, scale <span class="op">=</span> <span class="st">'free'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, fill <span class="op">=</span> <span class="st">"lightsalmon"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_errorbar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>ymin<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">mean</span><span class="op">-</span><span class="va">sd</span><span class="op">&gt;</span><span class="fl">0</span>,<span class="va">mean</span><span class="op">-</span><span class="va">sd</span>,<span class="fl">0</span><span class="op">)</span>, ymax<span class="op">=</span><span class="va">mean</span><span class="op">+</span><span class="va">sd</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">.2</span>, position<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge</a></span><span class="op">(</span><span class="fl">.9</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">omicsEye_theme</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span> <span class="op">(</span>strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'Inclusion proportion'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span></span></code></pre></div>
<p>We also illustrate the estimated IntegratedLearner layer weights. We
observe that the layers with highest single-omics predictive accuracy:
microbiome, metabolomics and PlasmaSomalogic are given the most weight
in the stacked model. Furthermore, we highlight the feature importance
scores of top 20 features of microbiome and PlasmaSomalogic layer which
highlights several features that agree with known biology.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p4</span>,</span>
<span>             ncol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>             labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Estimated IntegratedLearner layer weights'</span><span class="op">)</span>,</span>
<span>             label_size <span class="op">=</span> <span class="fl">8</span>, vjust <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntegratedLearner_files/figure-html/unnamed-chunk-11-1.png" class="r-plt" width="672"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p5</span>,</span>
<span>             ncol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>             labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Top 20 features of Microbiome and PlasamaSomalogic layer'</span><span class="op">)</span>,</span>
<span>             label_size <span class="op">=</span> <span class="fl">8</span>, vjust <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntegratedLearner_files/figure-html/unnamed-chunk-12-1.png" class="r-plt" width="864"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R Under development (unstable) (2025-11-06 r88990)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8          LC_NUMERIC=C                 </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8           LC_COLLATE=en_US.UTF-8       </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8       LC_MESSAGES=en_US.UTF-8      </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8          LC_NAME=en_US.UTF-8          </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=en_US.UTF-8        LC_TELEPHONE=en_US.UTF-8     </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8    LC_IDENTIFICATION=en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] splines   stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] bartMachine_1.3.4.1     missForest_1.6.1        randomForest_4.7-1.2   </span></span>
<span><span class="co">#&gt;  [4] bartMachineJARs_1.2.1   rJava_1.0-11            bayesplot_1.14.0       </span></span>
<span><span class="co">#&gt;  [7] cowplot_1.2.0           caret_7.0-1             lattice_0.22-7         </span></span>
<span><span class="co">#&gt; [10] SuperLearner_2.0-29     gam_1.22-6              foreach_1.5.2          </span></span>
<span><span class="co">#&gt; [13] nnls_1.6                lubridate_1.9.4         forcats_1.0.1          </span></span>
<span><span class="co">#&gt; [16] stringr_1.6.0           dplyr_1.1.4             purrr_1.2.0            </span></span>
<span><span class="co">#&gt; [19] readr_2.1.5             tidyr_1.3.1             tibble_3.3.0           </span></span>
<span><span class="co">#&gt; [22] ggplot2_4.0.0           tidyverse_2.0.0         IntegratedLearner_0.0.2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] Rdpack_2.6.4         pROC_1.19.0.1        rlang_1.1.6         </span></span>
<span><span class="co">#&gt;  [4] magrittr_2.0.4       prediction_0.3.18    compiler_4.6.0      </span></span>
<span><span class="co">#&gt;  [7] systemfonts_1.3.1    vctrs_0.6.5          reshape2_1.4.5      </span></span>
<span><span class="co">#&gt; [10] quadprog_1.5-8       crayon_1.5.3         pkgconfig_2.0.3     </span></span>
<span><span class="co">#&gt; [13] shape_1.4.6.1        fastmap_1.2.0        backports_1.5.0     </span></span>
<span><span class="co">#&gt; [16] labeling_0.4.3       rmarkdown_2.30       tzdb_0.5.0          </span></span>
<span><span class="co">#&gt; [19] prodlim_2025.04.28   nloptr_2.2.1         itertools_0.1-3     </span></span>
<span><span class="co">#&gt; [22] ragg_1.5.0           glmnetUtils_1.1.9    xfun_0.54           </span></span>
<span><span class="co">#&gt; [25] glmnet_4.1-10        cachem_1.1.0         jsonlite_2.0.0      </span></span>
<span><span class="co">#&gt; [28] recipes_1.3.1        parallel_4.6.0       R6_2.6.1            </span></span>
<span><span class="co">#&gt; [31] bslib_0.9.0          stringi_1.8.7        RColorBrewer_1.1-3  </span></span>
<span><span class="co">#&gt; [34] ranger_0.17.0        parallelly_1.45.1    rpart_4.1.24        </span></span>
<span><span class="co">#&gt; [37] jquerylib_0.1.4      Rcpp_1.1.0           iterators_1.0.14    </span></span>
<span><span class="co">#&gt; [40] knitr_1.50           future.apply_1.20.0  Matrix_1.7-4        </span></span>
<span><span class="co">#&gt; [43] nnet_7.3-20          timechange_0.3.0     tidyselect_1.2.1    </span></span>
<span><span class="co">#&gt; [46] abind_1.4-8          yaml_2.3.10          timeDate_4051.111   </span></span>
<span><span class="co">#&gt; [49] codetools_0.2-20     listenv_0.10.0       doRNG_1.8.6.2       </span></span>
<span><span class="co">#&gt; [52] plyr_1.8.9           withr_3.0.2          S7_0.2.0            </span></span>
<span><span class="co">#&gt; [55] posterior_1.6.1      ROCR_1.0-11          evaluate_1.0.5      </span></span>
<span><span class="co">#&gt; [58] future_1.67.0        desc_1.4.3           survival_3.8-3      </span></span>
<span><span class="co">#&gt; [61] pillar_1.11.1        tensorA_0.36.2.1     rngtools_1.5.2      </span></span>
<span><span class="co">#&gt; [64] checkmate_2.3.3      stats4_4.6.0         insight_1.4.2       </span></span>
<span><span class="co">#&gt; [67] distributional_0.5.0 generics_0.1.4       hms_1.1.4           </span></span>
<span><span class="co">#&gt; [70] scales_1.4.0         globals_0.18.0       class_7.3-23        </span></span>
<span><span class="co">#&gt; [73] glue_1.8.0           tools_4.6.0          data.table_1.17.8   </span></span>
<span><span class="co">#&gt; [76] ModelMetrics_1.2.2.2 gower_1.0.2          fs_1.6.6            </span></span>
<span><span class="co">#&gt; [79] grid_4.6.0           rbibutils_2.4        ipred_0.9-15        </span></span>
<span><span class="co">#&gt; [82] nlme_3.1-168         performance_0.15.2   cli_3.6.5           </span></span>
<span><span class="co">#&gt; [85] textshaping_1.0.4    lava_1.8.2           gtable_0.3.6        </span></span>
<span><span class="co">#&gt; [88] sass_0.4.10          digest_0.6.38        htmlwidgets_1.6.4   </span></span>
<span><span class="co">#&gt; [91] farver_2.1.2         htmltools_0.5.8.1    pkgdown_2.2.0       </span></span>
<span><span class="co">#&gt; [94] lifecycle_1.0.4      hardhat_1.4.2        MASS_7.3-65</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Franzosa EA et al. (2019). <a href="https://www.ncbi.nlm.nih.gov/pubmed/30531976" class="external-link">Gut microbiome
structure and metabolic activity in inflammatory bowel disease</a>.
<em>Nature Microbiology</em> 4(2):293–305.</p>
<p>Ghaemi MS et al. (2019). <a href="https://pubmed.ncbi.nlm.nih.gov/30561547/" class="external-link">Multiomics modeling of
the immunome, transcriptome, microbiome, proteome and metabolome
adaptations during human pregnancy</a>. <em>Bioinformatics</em>
35(1):95-103.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Mallick et al. (2024). <a href="https://pubmed.ncbi.nlm.nih.gov/38146838/" class="external-link">An integrated Bayesian
framework for multi-omics prediction and classification</a>.
<em>Statistics in Medicine</em> 43(5):983–1002.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Himel Mallick, Anupreet Porwal.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
